### PR TITLE
gh-100344: Provide C implementation for asyncio.current_task

### PR DESCRIPTION
**Pull request by itamaro**
Implementing it in C makes it about 4x-6x faster

Microbenchmark:

```
# bench.py

import time
import timeit
import asyncio

ITERS: int = 10**6
NANO: int = 10**9
NANO_PER_ITER: float = NANO / ITERS

async def main():
   # avoid load attr noise
   py_current_task = asyncio.tasks._py_current_task
   c_current_task = asyncio.tasks._c_current_task

   asyncio.current_task() # warmup
   py_current_task() # warmup
   c_current_task() # warmup

   print(
      "current_task: {}ns".format(timeit.timeit(py_current_task, number=ITERS, timer=time.process_time) * NANO_PER_ITER)
   )
   print(
      "current_task: {}ns".format(timeit.timeit(c_current_task, number=ITERS, timer=time.process_time) * NANO_PER_ITER)
   )

asyncio.run(main())
```

a few runs on MacBook Pro
2.4 GHz 8-Core Intel Core i9
64 GB 2667 MHz DDR4:

debug build:

```
~/work/pyexe/main-dbg ⌚ 9:57:34
$ ./python.exe bench.py
[py] current_task: 606.234ns
[c] current_task: 104.47699999999993ns

~/work/pyexe/main-dbg ⌚ 9:57:59
$ ./python.exe bench.py
[py] current_task: 631.856ns
[c] current_task: 110.22500000000002ns

~/work/pyexe/main-dbg ⌚ 9:58:08
$ ./python.exe bench.py
[py] current_task: 637.746ns
[c] current_task: 105.03899999999999ns

~/work/pyexe/main-dbg ⌚ 9:58:16
$ ./python.exe bench.py
[py] current_task: 621.3169999999999ns
[c] current_task: 103.01300000000002ns
```

opt build:

```
~/work/pyexe/main-opt ⌚ 10:33:17
$ ./python.exe bench.py
[py] current_task: 128.743ns
[c] current_task: 31.997999999999998ns

~/work/pyexe/main-opt ⌚ 10:33:24
$ ./python.exe bench.py
[py] current_task: 126.388ns
[c] current_task: 32.64599999999998ns

~/work/pyexe/main-opt ⌚ 10:33:26
$ ./python.exe bench.py
[py] current_task: 137.053ns
[c] current_task: 32.066999999999986ns

~/work/pyexe/main-opt ⌚ 10:33:28
$ ./python.exe bench.py
[py] current_task: 131.17800000000003ns
[c] current_task: 32.06600000000001ns
```

Co-authored-by: @pranavtbhat

* Issue: gh-100344

### CODE REVIEW

**Comment by pablogsal** [comment]
Congrats @itamaro! Your PR has been selected to try the new refleak buildbot label! 😉

**Comment by itamaro** [comment]
> Your PR has been selected to try the new refleak buildbot label! 😉

Such an honor, thank you @pablogsal 😛 

Checking out a few of the additional refleak test runs, I saw (across different buildbots): timeout, failure of `test_nntplib`, and failure of `test_logging`.
While none of these indicates new refleaks nor seems obviously related to this PR, it did nudge me to take a closer look, and I did find refleaks using:

```
./python.exe -m test test_asyncio.test_tasks -v -m "*current_task*" -R 3:3
...
OK
.
test_asyncio.test_tasks leaked [2990, 2988, 2990] references, sum=8968
test_asyncio.test_tasks leaked [896, 894, 894] memory blocks, sum=2684
test_asyncio.test_tasks failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_asyncio.test_tasks

Total duration: 23.6 sec
Tests result: FAILURE
```

I think the issue is not decrefing the `loop` when getting it from `_asyncio_get_running_loop_impl`. I'm pushing a fix that made it pass for me locally - let's see what the buildbots think 🤞

**Comment by gvanrossum** [comment]
I'm waiting for the benchmark test suggested by Mark in the issue.

**Comment by itamaro** [comment]
> I'm waiting for the benchmark test suggested by Mark in the issue.

replied on the issue and pushed the python optimization @markshannon suggested for ~40% speedup.
my suggestion is to do both optimizations 😄

File: Lib/asyncio/tasks.py
Comment (carljm) [inline Lib/asyncio/tasks.py]: It _shouldn't_ matter, but I think preferably we would not blanket silence all exceptions here.

```suggestion
    except KeyError:
```

**Review by carljm — CHANGES_REQUESTED** [review]

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
The tests need to be adjusted to test both pure python and c implementation. Also looks like there are merge conflicts.

**Comment by itamaro** [comment]
I have made the requested changes; please review again

**Review by kumaraditya303 — COMMENTED** [review]
LGTM, I'll wait for @gvanrossum's review before merging.

**Review by gvanrossum — APPROVED** [review]
@kumaraditya303 If you agree go ahead and merge.

**Comment by itamaro** [comment]
Is it still ok to push another change to revert the Python implementation as discussed on the issue, after approval?

**Comment by kumaraditya303** [comment]
> Is it still ok to push another change to revert the Python implementation as discussed on the issue, after approval?

That's fine, I'll review anyways before merging so go ahead!

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by kumaraditya303** [comment]
Merged, thanks for working on this!

**Pull request merged**
